From 7d068ebf416b17817752e1f05fe3c9c5d9804408 Mon Sep 17 00:00:00 2001
From: hariharasudhan Rajkumar <hrajkumar@maxlinear.com>
Date: Thu, 5 Sep 2024 13:15:28 +0800
Subject: [PATCH] WLANRTSYS-83122 hostapd crashed after inband radar detection
 in AP

Issue: Hostapd crash observed when the inband ETSI radar hits.

Root cause: when the radar hits hostapd chooses channel 140
as a fallback channel, drivers reject it as unavailable channel.
which results in a interface down, after that the hapd structure
is accessed which results in a crash. Because in case of ETSI
channel 132-144 does not support 80MHz BW as per the regulatory rule.

Fix: reg rule updation code changes were missed during hostap-ng
migration, now added.

FIXUP to WLANRTSYS-64810 ZWDFS changes for wave700 in hostap-ng
---
 src/drivers/driver.h              |  7 +++++++
 src/drivers/driver_nl80211_capa.c | 11 +++++++++++
 2 files changed, 18 insertions(+)

diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index bdbb2b7c9..80014685a 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -52,6 +52,13 @@
 #define HOSTAPD_CHAN_INDOOR_ONLY 0x00010000
 #define HOSTAPD_CHAN_GO_CONCURRENT 0x00020000
 
+#ifdef CONFIG_VENDOR_MXL
+#define HOSTAPD_CHAN_NO_HT40_MINUS 0x00040000
+#define HOSTAPD_CHAN_NO_HT40_PLUS  0x00080000
+#define HOSTAPD_CHAN_NO_80MHZ   0x20000000
+#define HOSTAPD_CHAN_NO_160MHZ  0x40000000
+#endif /* CONFIG_VENDOR_MXL */
+
 /* Allowed bandwidth mask */
 enum hostapd_chan_width_attr {
 	HOSTAPD_CHAN_WIDTH_10   = BIT(0),
diff --git a/src/drivers/driver_nl80211_capa.c b/src/drivers/driver_nl80211_capa.c
index 1d7b4f194..07a339228 100644
--- a/src/drivers/driver_nl80211_capa.c
+++ b/src/drivers/driver_nl80211_capa.c
@@ -1788,6 +1788,17 @@ static void phy_info_freq(struct hostapd_hw_modes *mode,
 			tb_freq[NL80211_FREQUENCY_ATTR_DFS_CAC_TIME]);
 	}
 
+#ifdef CONFIG_VENDOR_MXL
+	if (tb_freq[NL80211_FREQUENCY_ATTR_NO_HT40_MINUS])
+		chan->flag |= HOSTAPD_CHAN_NO_HT40_MINUS;
+	if (tb_freq[NL80211_FREQUENCY_ATTR_NO_HT40_PLUS])
+		chan->flag |= HOSTAPD_CHAN_NO_HT40_PLUS;
+	if (tb_freq[NL80211_FREQUENCY_ATTR_NO_80MHZ])
+		chan->flag |= HOSTAPD_CHAN_NO_80MHZ;
+	if (tb_freq[NL80211_FREQUENCY_ATTR_NO_160MHZ])
+		chan->flag |= HOSTAPD_CHAN_NO_160MHZ;
+#endif
+
 	chan->wmm_rules_valid = 0;
 	if (tb_freq[NL80211_FREQUENCY_ATTR_WMM]) {
 		static struct nla_policy wmm_policy[NL80211_WMMR_MAX + 1] = {
-- 
2.43.0

